[API Compatibilty] Remove useless attn_mask expand and support select fa when attn_mask is not None for SDPA

LittleHeroZZZX opened the pull request:
<!-- TemplateReference: https://github.com/PaddlePaddle/Paddle/wiki/PULL-REQUEST-TEMPLATE--REFERENCE -->
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->

### PR Category
<!-- One of [ User Experience | Execute Infrastructure | Operator Mechanism | CINN | Custom Device | Performance Optimization | Distributed Strategy | Parameter Server | Communication Library | Auto Parallel | Inference | Environment Adaptation ] -->
User Experience

### PR Types
<!-- One of [ New features | Bug fixes | Improvements | Performance | BC Breaking | Deprecations | Docs | Devs | Not User Facing | Security | Others ] -->
Bug fixes

### Description
<!-- Describe what you’ve done -->
- 当 attn_mask 为 None 时，支持选择 flash_attention 作为后端
- 当 attn_mask dim shape 不为 `[bs, num_heads, seq_len, seq_len]`时：
  - FA expand 为四维，且只对 bs 维度进行广播，不对 num_heads 广播
  - mem efficient 不进行 expand，除非 attn_mask 为四维张量，并且 bs 维度不需要广播，但是 num_heads 维度需要广播，这种情况 kernel 无法自动广播，需要在传入前处理
  - math 后端不手动 expand，依赖计算过程中的自动广播

LittleHeroZZZX committed:
feat(sdpa): support attention mask broadcasting for memory efficient attention

- Remove `check_for_attn_mask` constraint to allow attention masks with flash attention
- Add mask broadcasting logic for flash attention when `attn_mask` is provided
- Extend mask broadcasting support for memory efficient attention to handle cases where batch size > 1 and mask heads != query heads
- Add unit tests to validate mask broadcasting behavior for memory efficient attention

zhwesky2010 commented on python/paddle/nn/functional/sdpa.py:
```diff
@@ -618,6 +602,12 @@ def scaled_dot_product_attention(
+            if attn_mask.ndim == 2:
+                attn_mask = attn_mask.expand([bs, 1, *attn_mask.shape])
+            elif attn_mask.ndim == 3:
+                attn_mask = paddle.unsqueeze(attn_mask, axis=1)
```
flash_attn支持直接处理2D mask吗

LittleHeroZZZX replied:
不可以，输入必须是四维的，并且 batch_size 这个维度需要和qkv 一致，否则会有异常；num_heads 这个维度支持广播

zhwesky2010 commented on python/paddle/nn/functional/sdpa.py:
```diff
@@ -642,6 +632,21 @@ def scaled_dot_product_attention(
         elif attn_mask is not None:
+            # memory_efficient_attention does not support broadcast num_heads dim when batch_size dim is not 1
+            if (
+                attn_mask is not None
```
这个分支就是attn_mask not None，不用这个条件

LittleHeroZZZX replied:
Removed

zhwesky2010 commented on python/paddle/nn/functional/sdpa.py:
```diff
@@ -642,6 +632,21 @@ def scaled_dot_product_attention(
+                and attn_mask.shape[0] != 1
+                and attn_mask.shape[1] != num_heads_q
+            ):
+                attn_mask = attn_mask.expand(
```
memory_efficient_attention支持直接处理2D mask吗，如果传入2D 或3D mask会怎么处理，需要复用上面逻辑吗

<img width="509" height="82" alt="infoflow 2025-12-18 16-12-24" src="https://github.com/user-attachments/assets/7da61bfd-dfdf-4c76-9e3b-d65dd7643e1f" />

LittleHeroZZZX replied:
flash attn 参考上面的代码，必须要 传入四维的 Tensor，但是允许是 num_heads 这个维度是 1

zhwesky2010 replied:
> flash attn 参考上面的代码，必须要 传入四维的 Tensor，但是允许是 num_heads 这个维度是 1

写错了，是memory_efficient_attention需要支持2D/3D/4D mask吗？
- 2D：是否需要复用上面fa的处理逻辑？
- 3D：是否需要复用上面fa的处理逻辑？
- 4D：目前会进行expand确保shape为[b, n, s, h]，不支持num_heads为1对吧

LittleHeroZZZX replied:
总的来说，memory_efficient_attention 支持传入任意的维度，包括：
- 2D，可以直接透传
- 3D，即`bs, seq_len, seq_len`，这里 bs 可以是 1或者真实的bs，可以直接透传
- 4D，即`bs, num_heads, seq_len, seq_len`，这里前两个维度可以是 1 或者真实值，可以直接透传。存在一个例外情况：当 bs != 1 并且 num_heads = 1时，memory_efficient_attention  不支持这种情况的自动广播，需要我们在调用前手动广播一下

zhwesky2010 commented on python/paddle/nn/functional/sdpa.py:
```diff
@@ -642,6 +632,21 @@ def scaled_dot_product_attention(
+                )
             bias_input = attn_mask
         else:
             bias_input = None
```
这个bias_input是不是直接就叫attn_mask就行了，是同一个概念吧

```
if is_causal:
    attn_mask = LowerTriangularMask
else attn_mask is not None:
    attn_mask = attn_mask.expand()

out = memory_efficient_attention(
            query,
            key,
            value,
            attn_bias=attn_mask,
            p=dropout_p,
            scale=scale,
            training=training,
        )
```

LittleHeroZZZX replied:
Done

LittleHeroZZZX committed:
feat(sdpa): simplify attn_mask handling and fix condition

- Remove redundant `attn_mask is not None` check in elif branch
- Rename `bias_input` to `attn_mask` for clarity when is_causal is True
- Directly pass `attn_mask` to memory_efficient_attention, eliminating intermediate variable
- Improves code readability and maintains same behavior

zhwesky2010 reviewed (approved):
LGTM
